SMOS l2 wind speed reader

TAlonglong opened the pull request:
Featture smos l2 wind speed reader

 - [x] Tests added <!-- for all bug fixes or enhancements -->
 - [x] Tests passed <!-- for all non-documentation changes -->
 - [x] Passes ``flake8 satpy`` <!-- remove if you did not edit any Python files -->
 - [ ] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->

TAlonglong commented:
Not sure this is needed https://github.com/pytroll/satpy/pull/1158/commits/d0d9f78cb51cc2936bdffbe09a14abce87c5fc9c#diff-22ba0a41beb5ea7c6d5d13c07a8a7b00R95

Is it possible to have two variables with the same name? Or am I missing something?

mraspaud commented on satpy/readers/smos_l2_wind.py:
```diff
@@ -0,0 +1,173 @@
+
+    def available_datasets(self, configured_datasets=None):
+        """Automatically determine datasets provided by this file."""
+        logger.debug("Available_datasets begin...")
```
Maybe this can be removed ?

TAlonglong replied:
You mean the debug message?

mraspaud replied:
yes

mraspaud commented on satpy/readers/smos_l2_wind.py:
```diff
@@ -0,0 +1,173 @@
+    def _mask_dataset(self, data):
+        """Mask out fill values ( and remove the _FillValue from attributes)."""
+        fill = data.attrs.pop('_FillValue')
+        return data.where(data != fill)
```
You can leave `_FillValue` in the attrs, but just replace it with `nan` (I suppose this is floats ?)

TAlonglong replied:
ok

mraspaud commented on satpy/readers/smos_l2_wind.py:
```diff
@@ -0,0 +1,173 @@
+        rename_dict = {}
+        if 'lon' in data.dims:
+            # Want lons range from -180 .. 180 ( not 0 .. 360)
+            data = data.assign_coords(lon=(((data.lon + 180) % 360) - 180))
```
This is also done in `_adjust_lon_interval`, maybe it can be factorized ?

TAlonglong replied:
Ok, will have a look.

TAlonglong commented:
Small discussion about the area_extent for the `get_area_def`

Assuming the given lat lon values for the grid is for the center of the pixel.

lat range is from -90 to 90 inclusive, 0.25 deg resolution giving 721 points
lon range is from 0 to 359.75 inclusive, 0.25 deg resolution giving 1440 points
( I translate the lon values to from -180 to 179.75)
What would the the area extent be? I just take the min and max.
-180, 90, 179.75, -90)

Is this correct?

mraspaud commented:
you have to go from ll corner to up corner. So if the ll pixel has lon1, lat1 as centre coordinate, the ll corner of the area should be lon1 - 0.125, lat1 - 0.125, and similar for the ur corner.

TAlonglong commented:
OK, but what about the lat values? It can't be -90.125 to 90.125?

mraspaud commented:
indeed, that would be strange... but maybe pyresample should allow it in this case.

pnuu commented:
If the coordinates are for the center of pixels, and the "top row" (northern-most) are at +90 degrees, then that would mean all data on that row should be the same. At the exact pole the longitudes are meaningless, so there's no sense in assigning that kind of coordinates.

TAlonglong commented:
Hm thanks @mraspaud @pnuu. I cant find anything in the documentation. Does CF-1.7 say something on the position on the pixel coordinates? Anyway I guess I need to ask the data provider.

TAlonglong commented:
OK, I talked with the data provider. the position is center of pixel.

How to I handle this for the lat values in the area extent? To make it mathematically correct it will be from lat -90.125 to 90.125. But that has no physical meaning. Any ideas?

mraspaud commented:
What happens if you use these values,  -90.125 and 90.125, in the area extent ? Pyresample is filtering out pixel with *centre* value outside [-90, 90] so maybe it does work ?

TAlonglong commented:
Will test later. meeting now.

pnuu commented:
So the data has N different values located at the poles? What does that mean in reality? :astonished: Either the data is incorrect (different values) or the coordinates are wrong (not really center of pixel).

mraspaud commented:
If all the values on first and last row are the same, that could still be valid...

TAlonglong commented:
This is SMOS wind data ( wind speed at 10m). So for the nearest future there will never be data at the poles. Its just a matter of getting the position right.

mraspaud commented:
Merging this for now, as it seems finished. If the geolocation isn't right at the poles, feel free to submit another PR

mraspaud commented on satpy/readers/smos_l2_wind.py:
```diff
@@ -0,0 +1,173 @@
+
+    def available_datasets(self, configured_datasets=None):
+        """Automatically determine datasets provided by this file."""
+        logger.debug("Available_datasets begin...")
```
```suggestion
```

mraspaud commented:
@TAlonglong where are you on this PR?

TAlonglong commented:
Oh, what happened yesterday? was I some kind of backlog day? I had the day of ( public Holiday i Norway) painting my house.

I remember talking to one of the creators of the product and doing some test. But was not happy with the result. I need to dig into this again. 

Man the time flies...

mraspaud commented:
No worries! We're planning on making a release soon, so we go through the open PR, trying to merge them, or pinging the contributors to finish them

TAlonglong commented:
OK

So I removed to top and bottom line in the datasets and adjusting the lat values accordingly. Then I think this will be correct.

Added some more test to check this trimming of data.

If the checking is successful this is ready for merge from my side.

TAlonglong commented:
Ok, codebeat is beating me.

Also with something from some other code i pulled into my code.

I give up for now.

mraspaud commented:
Thanks for the effort! The codebeat issues now are mild I think, I'm good with them being as they are.

mraspaud commented:
Travis passed but didn't notify github, merging!
